### PR TITLE
route_patterns endpoint, phase 2

### PR DESCRIPTION
**Pull request by lboyarsky**
Ticket: [route_patterns endpoint, phase 2](https://app.asana.com/0/1113863852319980/1113893353356229)

1. Add route_pattern_id to trips (has one)
```bash
$ curl -s 'http://localhost:4000/trips?id=Boat-F1-2315-Long-Saturday' | jsonpp
{
  "data": [
    {
      "attributes": {
        "bikes_allowed": 1,
        "block_id": "B-22",
        "direction_id": 0,
        "headsign": "Hingham via Hull",
        "name": "",
        "wheelchair_accessible": 1
      },
      "id": "Boat-F1-2315-Long-Saturday",
      "links": {
        "self": "/trips/Boat-F1-2315-Long-Saturday"
      },
      "relationships": {
        "route": {
          "data": {
            "id": "Boat-F1",
            "type": "route"
          }
        },
        "route_pattern": {
          "data": {
            "id": "Boat-F1-2-0",
            "type": "route_pattern"
          }
        },
        "service": {
          "data": {
            "id": "Boat-F1-Saturday",
            "type": "service"
          }
        },
        "shape": {
          "data": {
            "id": "b0001",
            "type": "shape"
          }
        }
      },
      "type": "trip"
    }
  ],
  "jsonapi": {
    "version": "1.0"
  }
}
```
2. /trips?filter[route_pattern]=
```bash
$ curl -s 'http://localhost:4000/trips?route_pattern=Boat-F1-8-1,Boat-F1-7-0' | jsonpp
{
  "data": [
    {
      "attributes": {
        "bikes_allowed": 1,
        "block_id": "B-01",
        "direction_id": 1,
        "headsign": "Boston-Long Wharf via Georges Island & Logan Airport",
        "name": "",
        "wheelchair_accessible": 1
      },
      "id": "Boat-F1-1000-Hingham-WeekdaySummer",
      "links": {
        "self": "/trips/Boat-F1-1000-Hingham-WeekdaySummer"
      },
      "relationships": {
        "route": {
          "data": {
            "id": "Boat-F1",
            "type": "route"
          }
        },
        "route_pattern": {
          "data": {
            "id": "Boat-F1-8-1",
            "type": "route_pattern"
          }
        },
        "service": {
          "data": {
            "id": "Boat-F1-WeekdaySummer",
            "type": "service"
          }
        },
        "shape": {
          "data": {
            "id": "b0020",
            "type": "shape"
          }
        }
      },
      "type": "trip"
    },
[...]
```
3. /routes?include=route_patterns (has many)
```bash
$ curl -s 'http://localhost:4000/routes?stop=place-cool&include=route_patterns' | jsonpp
{
[...]
  "included": [
    {
      "attributes": {
        "direction_id": 0,
        "name": "Brighton Center via Brookline",
        "sort_order": 50660096,
        "time_desc": "School days only",
        "typicality": 3
      },
      "id": "66-5-0",
      "links": {
        "self": "/route-patterns/66-5-0"
      },
      "relationships": {
        "representative_trip": {
          "data": {
            "id": "39984683",
            "type": "trip"
          }
        },
        "route": {
          "data": {
            "id": "66",
            "type": "route"
          }
        }
      },
      "type": "route_pattern"
    },
    [...]
```

4. /route-patterns?filter[direction_id]= (only in combination with route)
```bash
$ curl -s 'http://localhost:4000/route-patterns?route=Red&direction_id=1' | jsonpp
{
  "data": [
    {
      "attributes": {
        "direction_id": 1,
        "name": "Alewife",
        "sort_order": 10010151,
        "time_desc": null,
        "typicality": 1
      },
      "id": "Red-1-1",
      "links": {
        "self": "/route-patterns/Red-1-1"
      },
      "relationships": {
        "representative_trip": {
          "data": {
            "id": "38747571-L",
            "type": "trip"
          }
        },
        "route": {
          "data": {
            "id": "Red",
            "type": "route"
          }
        }
      },
      "type": "route_pattern"
    },
    {
      "attributes": {
        "direction_id": 1,
        "name": "Alewife",
        "sort_order": 10010152,
        "time_desc": null,
        "typicality": 1
      },
      "id": "Red-3-1",
      "links": {
        "self": "/route-patterns/Red-3-1"
      },
      "relationships": {
        "representative_trip": {
          "data": {
            "id": "38747835-L",
            "type": "trip"
          }
        },
        "route": {
          "data": {
            "id": "Red",
            "type": "route"
          }
        }
      },
      "type": "route_pattern"
    }
  ],
  "jsonapi": {
    "version": "1.0"
  }
}
```

### CODE REVIEW

**Comment by lboyarsky** [comment]
@skyqrose you're right, thanks, I've updated the PR description to show `included` block. As for `/trips/{id}?include=route_pattern`, I haven't implemented this as it was not requested in the ticket.

File: apps/parse/lib/parse/trips.ex [hunk 1]
```diff
@@ -39,7 +40,8 @@ defmodule Parse.Trips do
       block_id: copy(row["block_id"]),
       wheelchair_accessible: String.to_integer(row["wheelchair_accessible"]),
       route_type: trip_route_type(row["trip_route_type"]),
-      bikes_allowed: bikes_allowed(row["bikes_allowed"])
+      bikes_allowed: bikes_allowed(row["bikes_allowed"]),
+      route_pattern_id: copy(row["route_pattern_id"])
```
[the remark this answers is not in the document]
Comment (lboyarsky) [reply apps/parse/lib/parse/trips.ex]: Fixed

File: apps/model/lib/model/trip.ex [hunk 2]
```diff
@@ -32,7 +33,8 @@ defmodule Model.Trip do
           block_id: String.t() | nil,
           wheelchair_accessible: 0..2,
           alternate_route: boolean | nil,
-          bikes_allowed: 0..2
+          bikes_allowed: 0..2,
+          route_pattern_id: Model.RoutePattern.id()
```
[the remark this answers is not in the document]
Comment (lboyarsky) [reply apps/model/lib/model/trip.ex]: Fixed

File: apps/parse/lib/parse/trips.ex [hunk 3]
```diff
@@ -25,6 +25,7 @@ defmodule Parse.Trips do
   * `"block_id"` - `Model.Trip.t` `block_id`
   * `"wheelchair_accessible"` - `Model.Trip.t` `wheelchair_accessible`
   * `"trip_route_type" - `Model.Route.route_type | nil`
+  * `"route_pattern_id" - `Model.RoutePattern.t`
```
[the remark this answers is not in the document]
Comment (lboyarsky) [reply apps/parse/lib/parse/trips.ex]: Fixed

File: apps/state/lib/state/route_pattern.ex [hunk 4]
```diff
@@ -29,6 +29,12 @@ defmodule State.RoutePattern do
     by_ids(ids)
   end
 
+  def filter_by(%{route_ids: route_ids, direction_id: direction_id}) do
+    route_ids
+    |> by_route_ids()
+    |> Enum.filter(&(&1.direction_id == direction_id))
+  end
```
[the remark this answers is not in the document]
Comment (lboyarsky) [reply apps/state/lib/state/route_pattern.ex]: `select` not `match`, right? If yes, then fixed.

File: apps/state/lib/state/trip.ex [hunk 5]
```diff
@@ -182,24 +182,28 @@ defmodule State.Trip do
     [new_trip | new_alternates]
   end
 
-  defp do_apply_filters(filters, trips \\ [])
+  defp do_apply_filters(filters, trips \\ all())
```
[the remark this answers is not in the document]
Comment (lboyarsky) [reply apps/state/lib/state/trip.ex]: Done

File: apps/state/lib/state/trip.ex [hunk 6]
```diff
@@ -182,35 +162,45 @@ defmodule State.Trip do
...
+      |> build_filters(:route_id, filters[:routes])
+      |> build_filters(:direction_id, filters[:direction_id])
+      |> build_filters(:route_pattern_id, filters[:route_patterns])
+      |> build_filters(:id, filters[:ids])
 
-    new_trips =
-      filtered_trips
-      |> List.flatten()
-      |> Enum.concat(trips)
+    trips = State.Trip.select(matchers) ++ State.Trip.Added.select(matchers)
```
Comment (paulswartz) [inline apps/state/lib/state/trip.ex]: These will be slower if an index isn't provided; maybe `build_filters` should take/return an index as well?

File: apps/state/lib/state/trip.ex [hunk 7]
```diff
@@ -182,35 +162,45 @@ defmodule State.Trip do
...
 
-  defp do_apply_filters(%{date: date = %Date{}} = filters, trips) do
-    new_trips = Enum.filter(trips, &ServiceByDate.valid?(&1.service_id, date))
+  defp build_filters(matchers, _key, nil), do: matchers
 
-    filters
-    |> Map.delete(:date)
-    |> do_apply_filters(new_trips)
+  defp build_filters(matchers, key, values) when is_list(values) do
+    if matchers == [] do
```
Comment (paulswartz) [inline apps/state/lib/state/trip.ex]: If you start with `matchers = [%{}]` when you kick off these calls, you don't need the extra `if` case.

File: apps/state/lib/state/trip.ex [hunk 8]
```diff
@@ -182,35 +162,45 @@ defmodule State.Trip do
...
-    new_trips =
-      filtered_trips
-      |> List.flatten()
-      |> Enum.concat(trips)
+    trips = State.Trip.select(matchers) ++ State.Trip.Added.select(matchers)
 
-    filters
-    |> Map.drop([:routes, :direction_id])
-    |> do_apply_filters(new_trips)
+    case filters[:date] do
```
Comment (paulswartz) [inline apps/state/lib/state/trip.ex]: `case Map.fetch(filters, :date)` (rather than using Access).

File: apps/state/lib/state/trip.ex (hunk 7 above)
Comment (paulswartz) [reply apps/state/lib/state/trip.ex]: You could also do
```elixir
values = List.wrap(values)
```
and then you don't need to handle the separate case where `values` is a single item.

File: apps/state/lib/state/trip.ex (hunk 6 above)
Comment (lboyarsky) [reply apps/state/lib/state/trip.ex]: Not sure I fully understand how those indexes work but please take a look at the updated version and let me know if this is what is needed.

File: apps/state/lib/state/trip.ex (hunk 8 above)
Comment (lboyarsky) [reply apps/state/lib/state/trip.ex]: Done

File: apps/state/lib/state/trip.ex [hunk 9]
```diff
@@ -162,45 +162,37 @@ defmodule State.Trip do
...
   end
 
-  defp build_filters(matchers, key, value) do
-    if matchers == [] do
-      [%{key => value}]
-    else
-      for matcher <- matchers, do: Map.put(matcher, key, value)
-    end
-  end
+  defp get_index(%{ids: ids}) when length(ids) > 0, do: :id
```
Comment (paulswartz) [inline apps/state/lib/state/trip.ex]: You don't need the full length, which walks the whole list. Instead, you can do `when ids != []`.

Comment (lboyarsky) [reply apps/state/lib/state/trip.ex]: Fixed

**Pull request merged**